### PR TITLE
Hierarchy Restored

### PR DESCRIPTION
**Pull request by XedinUnknown**
Aims to incorporate solution for #3.

### CODE REVIEW

**Comment by mecha** [comment]
@XedinUnknown Consider adding an exception that is aware of both a `RenderCapableInterface` and a `ContextInterface`. Suggestion: `RendererExceptionInterface`

**Comment by XedinUnknown** [comment]
You mean like the [`ContextAwareRenderExceptionInterface`]?

[`ContextAwareRenderExceptionInterface`]:    https://github.com/Dhii/output-renderer-interface/blob/task/3-restore-hierarchy/src/Exception/ContextAwareRenderExceptionInterface.php

**Comment by mecha** [comment]
I guess - it's the equivalent in terms of ancestors, but `ContextAwareRenderExceptionInterface` is specific to render failure, since it extends `CouldNotRenderExceptionInterface`. Which makes sense - any problem specific to context is a failure to render, hence "could not render".

In general I think the naming of exception interfaces is a bit weird. Consider this for a moment:

```php
// Currently called RenderCapableAwareExceptionInterface
// Applicable to any error encountered by a "render capable" instance
interface RenderExceptionInterface extends
    ThrowableInterface,
    RenderCapableAwareInterface
{
}

// Applicable to render failure - any problem encountered during rendering
interface CouldNotRenderExceptionInterface extends RenderExceptionInterface
{
}

// Applicable to render failure related to context
// Any problem encountered during rendering
interface CouldNotRenderContextExceptionInterface extends
    CouldNotRenderExceptionInterface,
    ContextAwareInterface
{
}
```

**Comment by XedinUnknown** [comment]
> any problem specific to context is a failure to render

Yes, @mecha, that was exactly my reasoning.

With regard to your suggested changes to naming, `RenderExceptionInterface` is an error encountered during rendering, according to your description. However, this role is filled by `CouldNotRenderExceptionInterface`. `RenderCapableAwareExceptionInterface`, on the other hand, can be thrown by some other methods, including the constructor, of the rendering object. If the base class was called `RendererInterface` instead of `RenderCapableInterface`, then this would fix such issues, together with names that are getting longer. However, it would break code that currently depends on it having a method with a context. Which may, or may not, be a big problem. Thoughts?

**Comment by mecha** [comment]
> With regard to your suggested changes to naming, RenderExceptionInterface is an error encountered during rendering, according to your description.

My mistake - a copy-paste error from my part. Changed to `Applicable to any error encountered by a "render capable" instance`.

I'm not sure I understood what you said about the base interface's name. I would leave the interfaces for render capable objects as they are.

File: src/ContextRendererInterface.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+{
+    /**
+     * Produce output based on context.
+     *
+     * @since 0.1
+     *
+     * @param mixed|null $context The context;
+     * something that can provide more information on how to perform rendering.
+     *
+     * @throws ContextAwareRenderExceptionInterface If cannot render.
```
Comment (mecha) [inline src/ContextRendererInterface.php]: Needs to be changed to `ContextRenderExceptionInterface`.

File: src/RendererInterface.php [hunk 2]
```diff
@@ -2,28 +2,22 @@
...
-     * @since 0.1
+     * Produces output.
      *
-     * @param mixed|null $context The context; something that can provide more information on how to perform rendering.
+     * @since [*next-version*]
      *
-     * @throws CouldNotRenderExceptionInterface If cannot render.
-     * @throws RendererExceptionInterface       Any other problem related to the renderer.
+     * @throws CouldNotRenderExceptionInterface  If cannot render.
+     * @throws RenderableAwareExceptionInterface Any other problem related to the renderer.
```
Comment (mecha) [inline src/RendererInterface.php]: Needs to be changed to `RendererExceptionInterface`.

**Review by mecha — CHANGES_REQUESTED** [review]
I love the new names. They are so much more intuitive and shorter to boot!

There are a few `throws` declarations that need changing. Other than that, this LGTM.

**Review by mecha — APPROVED** [review]

**Pull request merged**